PIPS/pip-1.md

```diff
@@ -0,0 +1,46 @@
+---
+pip: 1
+title: Permissionless LockProxies
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-23
+---
+
+## Simple Summary
+The current LockProxy implementation requires an operator to control `setManagerProxy`, `bindProxyHash` and `bindAssetHash`.
+
+This PIP proposes an alternative implementation which does not require an operator.
+
+## Specification
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The `lock` function can be modified to require the following parameters:
+- address fromAssetHash
+- uint64 toChainId
+- bytes memory targetProxyHash
+- bytes memory toAssetHash
+- bytes memory toAddress
+- uint256 amount
+
+The LockProxy contract should maintain a `balances` mapping of (bytes32 => uint256).
+This `balances` mapping should be updated within the `lock` function:
+```
+key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash)
+balances[key] += amount
+```
+
+Where `hash` is any supported hashing function of the respective blockchain.
+
+The `unlock` function should be modified to require the following parameters:
+- bytes memory argsBs
+
+Where `argsBs` contains `fromAssetHash`, `fromChainId`, `fromProxyHash` and `toAssetHash`, `toAddress` and `amount`.
+
+The contract should check that the balances are sufficient, then reduce the balance and perform the unlock:
+```
+key = hash(toAssetHash, fromChainId, fromProxyHash, fromAssetHash)
+require(balances[key] >= amount)
+balances[key] -= amount
```

Comment:
if i transfer NEO from NEO chain to Ontology, then i transfer NEO from Ontology to Ethereum, but i can't transfer NEO from Ethereum to NEO, because the balances is zore.

Reply:
Oh right, I didn't think about that case. Maybe there could be two types of LockProxies?

Because for Switcheo we don't need to have the transfer flow NEO => Ontology => Ethereum => NEO, we only need NEO => Switcheo Chain, Switcheo Chain => NEO

A permissionless LockProxy would be very useful for us because we want to allow anyone to deposit and list tokens if they pay some SWTH as fees.

Reply:
ok, so in this situation, this proposal just can support at most two chain to transfer asset, i think it need explain this limit.

Reply:
I also realised that my proposal doesn't work with the token specification, e.g. https://github.com/skyinglyh1/solidity-crosschain-contracts/blob/master/contracts/core/v2.0/assets/ont/ONTX.sol.

Since anyone would be able to `unlock` any token by calling `lock` with any token they want.

For this proposal to make sense, I think the template ONTX.sol can be modified to be something like:
```
    constructor (address lockProxyContractAddress, uint64 fromChainId, bytes memory fromLockProxy, bytes memory fromAssetHash) public ERC20Detailed("ONT Token", "ONTX", 0) {
        _mint(lockProxyContractAddress, 1000000000);
       lockProxy.delegateAsset(fromChainId, fromLockProxy, fromAssetHash);
    }
```

In LockProxy:
```
struct DelegatedAsset {
    uint64 fromChainId;
    bytes fromLockProxy;
    bytes fromAssetHash;
}

mapping(address => DelegatedAsset) delegatedAssets;

function delegateAsset(uint64 chainId, bytes memory fromLockProxy, bytes memory fromAssetHash) public {
    delegatedAssets[_msgSender()] = DelegatedAsset(fromChainId, fromLockProxy, fromAssetHash);
}

function unlock(bytes memory argsBs) public {
    TxArgs memory args = _deserializTxArgs(argsBs);
    DelegatedAsset dAsset = delegatedAssets[args.toAssetHash];
    if (dAsset.fromChainId != 0) {
        require(dAsset.fromChainId == args.fromChainId && dAsset.fromLockProxy == args.fromContractAddr && dAsset.fromAssetHash == args.fromAssetHash);
    }
}
```

Reply:
> ok, so in this situation, this proposal just can support at most two chain to transfer asset, i think it need explain this limit.

yup makes sense, I will add an explanation for this limit